src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -324,6 +345,7 @@ public Parser(Compilation compilation, in JsonSourceGenerationContext sourceGene
                     // Clear the cache of generated metadata between the processing of context classes.
                     _typeGenerationSpecCache.Clear();
                     _implicitlyRegisteredTypes.Clear();
+                    _typeLevelDiagnostics.Clear();
```

Comment:
This is transient state scope to this particular method, so I would recommend adding a few debug-level asserts at the top of the method checking that these collections are empty: just to make sure there's no leftover items leaking into this method call.